File: src/main/java/com/timyr_tm/rust_bound/world/item/WireCoilItem.java
package com.timyr_tm.rust_bound.world.item;

import com.timyr_tm.rust_bound.core.component.DataComponents;
import com.timyr_tm.rust_bound.world.block.entity.ConnectableBlockEntity;
import com.timyr_tm.rust_bound.world.block.entity.ConnectionInfo;
import com.timyr_tm.rust_bound.world.block.entity.ConnectionPointInfo;
import net.minecraft.network.chat.Component;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.item.component.TooltipDisplay;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.phys.Vec3;
import org.jspecify.annotations.NonNull;

import java.util.Map;
import java.util.Optional;
import java.util.function.Consumer;

public class WireCoilItem extends Item {
    public WireCoilItem(Properties properties) {
        super(properties);
    }

    @Override
    public @NonNull InteractionResult useOn(@NonNull UseOnContext context) {
        if (context.getLevel().getBlockEntity(context.getClickedPos()) instanceof ConnectableBlockEntity blockEntity) {
            final Vec3 pos = context.getClickLocation().subtract(new Vec3(blockEntity.getBlockPos()));
            final Optional<Map.Entry<String, ConnectionPointInfo>> point = blockEntity.getConnectionPoints().entrySet().stream()
                .filter(
                    info -> info.getValue().shape().toAabbs().stream()
                        .anyMatch(
                            aabb -> aabb.inflate(.001).contains(pos)
                        )
                )
                .findFirst();
            if (point.isEmpty())
                return InteractionResult.FAIL;

            final ItemStack stack = context.getItemInHand();

            if (stack.has(DataComponents.WIRE_COIL_CONNECTION)) {
                final ConnectionInfo firstConnection = stack.remove(DataComponents.WIRE_COIL_CONNECTION);
                final ConnectionInfo lastConnection = new ConnectionInfo(point.orElseThrow().getKey(), point.orElseThrow().getValue().pos());

                final ConnectableBlockEntity lastBlockEntity = firstConnection.getBlockEntity(context.getLevel());

                if (lastBlockEntity == null)
                    return InteractionResult.FAIL;

                blockEntity.addConnection(lastConnection.name(), firstConnection);
                lastBlockEntity.addConnection(firstConnection.name(), lastConnection);
                blockEntity.setChanged();
                lastBlockEntity.setChanged();

                if (context.getLevel().isClientSide() && context.getPlayer() != null)
                    context.getPlayer().displayClientMessage(
                        Component.literal(
                            "%s (%s) - %s (%s)".formatted(
                                firstConnection.name(), firstConnection.pos(),
                                lastConnection.name(), lastConnection.pos()
                            )
                        ),
                        true
                    );
            }
            else {
                stack.set(DataComponents.WIRE_COIL_CONNECTION, new ConnectionInfo(point.orElseThrow().getKey(), point.orElseThrow().getValue().pos()));
                if (context.getLevel().isClientSide() && context.getPlayer() != null)
                    context.getPlayer().displayClientMessage(Component.literal("+ %s".formatted(point.orElseThrow().getKey())), true);
            }
            return InteractionResult.SUCCESS;
        }
        return super.useOn(context);
    }

    @Override
    @SuppressWarnings("deprecation")
    public void appendHoverText(
        @NonNull ItemStack stack, @NonNull TooltipContext context, @NonNull TooltipDisplay tooltipDisplay,
        @NonNull Consumer<Component> tooltipAdder, @NonNull TooltipFlag flag
    ) {

    }
}

File: src/main/kotlin/com/timyr_tm/rust_bound/world/item/WireCoilItem.kt
package com.timyr_tm.rust_bound.world.item;

import com.timyr_tm.rust_bound.core.component.DataComponents
import com.timyr_tm.rust_bound.world.block.entity.ConnectableBlockEntity
import com.timyr_tm.rust_bound.world.electricity.ConnectionInfo
import com.timyr_tm.rust_bound.world.electricity.ConnectionPointInfo
import com.timyr_tm.rust_bound.world.electricity.WireType
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceKey
import net.minecraft.world.InteractionResult
import net.minecraft.world.item.Item
import net.minecraft.world.item.context.UseOnContext
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.phys.Vec3

class WireCoilItem: Item {
    val wireType: ResourceKey<WireType>

    constructor(wireType: ResourceKey<WireType>, properties: Properties): super(properties) {
        this.wireType = wireType;
    }

    override fun useOn(context: UseOnContext): InteractionResult {
        val blockEntity: BlockEntity? = context.level.getBlockEntity(context.clickedPos)
        if (blockEntity is ConnectableBlockEntity) {
            val pos: Vec3 = context.clickLocation.subtract(Vec3(blockEntity.blockPos))
            val point: Map.Entry<String, ConnectionPointInfo>? = blockEntity.getConnectionPoints().entries.firstOrNull {
                info -> info.value.shape.toAabbs().any {
                    aabb -> aabb.inflate(0.001).contains(pos)
                }
            }
            if (point == null)
                return InteractionResult.FAIL
            if (context.itemInHand.has(DataComponents.WIRE_COIL_CONNECTION)) {
                val firstConnection: ConnectionInfo = context.itemInHand.remove(DataComponents.WIRE_COIL_CONNECTION)!!
                val lastConnection: ConnectionInfo = ConnectionInfo(point.key, point.value.pos, wireType)

                val lastBlockEntity: ConnectableBlockEntity = firstConnection.getBlockEntity(context.level)
                    ?: return InteractionResult.FAIL

                blockEntity.addConnection(lastConnection.name, firstConnection)
                lastBlockEntity.addConnection(firstConnection.name, lastConnection)
                blockEntity.setChanged()
                lastBlockEntity.setChanged()

                if (context.level.isClientSide && context.player != null)
                    context.player!!.displayClientMessage(
                        Component.literal(
                            "${firstConnection.name} (${firstConnection.pos}) - ${lastConnection.name} (${lastConnection.pos})"
                        ),
                        true
                    )
            }
            else {
                context.itemInHand.set(DataComponents.WIRE_COIL_CONNECTION, ConnectionInfo(point.key, point.value.pos, wireType))
                if (context.level.isClientSide && context.player != null)
                    context.player!!.displayClientMessage(Component.literal("+ ${point.key}"), true)
            }
            return InteractionResult.SUCCESS
        }
        return super.useOn(context)
    }
}
